[BLOG] Add initial blog post for OSD development on VS code (#3507)
opensearch-project/project-website: _posts/2024-12-17-vscode-osd-setup.markdown

---
layout: post
title:  "Streamline OpenSearch Dashboards development with VS Code"
authors:
 - wronghuy
 - kolchfa
date: 2024-12-17
categories:
 - technical-post
meta_keywords: opensearch dashboards development, vs code, jest

> meta_description: Learn how to setup OpenSearch Dashboards with VS Code

Reviewer: Please update the meta with the following: 

_meta_keywords:_ OpenSearch Dashboards, VS Code, development tools, Jest integration, ESLint, multi-root workspaces, unit testing, OSD codebase

_meta_description:_ Boost OpenSearch Dashboards development with VS Code. Learn to set up efficient workflows for testing, linting, and debugging. Enhance productivity using Jest, ESLint, and multi-root workspaces.

}
```

### Setup

1. Install the Prettier and ESLint extensions.
2. Press `Cmd` (or `Ctrl`) + `Shift` + `P` and select **Open Workspace Settings (JSON)**.
3. Add in the following settings to the `settings.json` file:

    ```json
    {
      "prettier.configPath": ".prettierrc",
      "editor.defaultFormatter": "esbenp.prettier-vscode",
      "eslint.autoFixOnSave": true,
      "editor.codeActionsOnSave": {
        "source.fixAll.eslint": true
      },
    }
    ```
**Note**: If you already have some settings in the `editor.codeActionsOnSave` setting, append `"source.fixAll.eslint": true` to the existing settings.

Now, when you save a file, the file is linted automatically.

## Configuring OSD server run tasks

Starting the OSD server typically requires running commands in two terminals:

* In the first terminal, run `yarn opensearch snapshot`.
* After several seconds, in the second terminal, run `yarn run start --no-base-path`.

This approach works for many use cases but doesn't allow you to debug server-side changes. To simplify this process, you can set up a series of launch tasks, turning server startup into a one-click operation:

1. Start the OpenSearch server.

> 2. Start the Dashboards server. Starting the Dashboards server requires a wait period for the server to boot up; this can be accomplished with a VS Code task to wait for a period of time.

Reviewer 2: ```suggestion
2. Start the Dashboards server. Starting the Dashboards server requires a wait period while the server boots up; this can be accomplished with a VS Code task that specifies to wait for a certain period of time.
```

> 

Reviewer 2: Line 111: Confirm that my additions reflect the intended meaning.

Making this task a VS Code run configuration provides the following benefits:

* One-click OSD development server startup, eliminating the need to repeatedly use the CLI.

> * Ability to debug server-side code (for client-side code (public), you can use your preferred browser's developer tools).

Reviewer 2: ```suggestion
* The ability to debug server-side code (for client-side code [public], you can use your preferred browser's developer tools).
```

      ],
      "compounds": [
      /**
      * This is the run configuration to startup OSD
      * 1. Starts up OpenSearch server
      * 2. Waits 13 seconds
      * 3. Starts up Dashboards server
      */
        {
          "name": "Start Dashboards",
          "configurations": ["Start OpenSearch Snapshot", "Start Dashboards Server"],
          // If either Dashboards or OpenSearch server is stopped, both servers will be stopped (set this to false to individually turn off a server without turning all off)
          "stopAll": true
        }
      ]
    }
    ```

1. If the `tasks.json` file does not already exist in the `.vscode` directory at the project root, create it and add the following configuration:

    ```json
    {
      "version": "2.0.0",
      "tasks": [
      // Silent task to sleep for 13 seconds; this is the upper limit on the time taken for OpenSearch to boot up
        {
          "label": "13 Second Delay Command",
          "type": "shell",
          "command": "sleep 13",
          "group": "none",
          "presentation": {
            "reveal": "silent",
            "panel": "new",
            "close": true
          }
        }
      ]
    }
    ```

> 1. In the **Run and Debug** tab, select **Start Dashboards** from the dropdown menu and press the play icon. OSD should start after a period of time.

Reviewer 2: ```suggestion
1. On the **Run and Debug** tab, select **Start Dashboards** from the dropdown menu and press the play icon. OSD should start after a period of time.
```

  
Now you can set any breakpoint in `server` code.

> The **Debug** toolbar should appear under the **CALL STACK**, in a specific worker node, as shown in the following image.

Reviewer 2: ```suggestion
The **Debug** toolbar should appear under the **CALL STACK** in a specific worker node, as shown in the following image.
```

![debug-toolbar](/assets/media/blog-images/2024-12-17-vscode-osd-setup/debug-toolbar.png)

> Using this toolbar, you can resume breakpoints, step into code, and execute the next instruction, as shown in.

Reviewer 2: ```suggestion
Using this toolbar, you can resume breakpoints, step into code, and execute the next instruction.
```

## Configuring multi-root workspace integration

> For most development scenarios, working within the OSD Core should be sufficient. However, if you need to develop code for a plugin, the [multi-root workspaces feature](https://code.visualstudio.com/docs/editor/multi-root-workspaces) is a useful option. The setup for Jest, linting, and run configurations for plugins is similar to that of the OSD Core, so we won’t cover it in detail here.

Reviewer 2: ```suggestion
For most development scenarios, working within the OSD Core should be sufficient. However, if you need to develop code for a plugin, the [multi-root workspaces feature](https://code.visualstudio.com/docs/editor/multi-root-workspaces) is a useful option. The setup for Jest, linting, and run configurations for plugins is similar to that of the OSD Core, so we won't cover it in detail here.
```

> In summary, for OSD to recognize plugins during development, the plugin's project root must be located within the `OpenSearch-Dashboards/plugins/` directory. The following image shows multi-root workspace integration.

Reviewer 2: ```suggestion
In summary, in order for OSD to recognize plugins during development, the plugin's project root must be located within the `OpenSearch-Dashboards/plugins/` directory. The following image shows multi-root workspace integration.
```

![multi-root-workspaces](/assets/media/blog-images/2024-12-17-vscode-osd-setup/multi-root-workspaces.gif)

### Setup

> In this example, assume you're developing code for the [anomaly-detection-dashboards-plugin](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin) and the project has been checked out in the `plugins` directory.

Reviewer 2: ```suggestion
In this example, assume you're developing code for the [anomaly-detection-dashboards-plugin](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin) and that the project has been checked out in the `plugins` directory.
```

> 1. Navigate one directory above the `OpenSearch-Dashboards` project folder and create a file called `OpenSearch-Dashboards.code-workspace`. Add the following configuration to this file:

Reviewer 2: ```suggestion
1. Navigate to one directory above the `OpenSearch-Dashboards` project folder and create a file called `OpenSearch-Dashboards.code-workspace`. Add the following configuration to this file:
```

    ```json
    {
      "folders": [
        {
          // Names are configurable; they will show up in the EXPLORER tab
          "name": "OSD Core",
        // Path to the project root
          "path": "OpenSearch-Dashboards"
        },
        {
          "name": "Anomaly Detection Plugin",
          "path": "OpenSearch-Dashboards/plugins/anomaly-detection-dashboards-plugin"
        }
      ],
      // Define workspace-specific settings here
      "settings": {},
      "launch": {
        // Specify workspace-specific launch configurations
        "configurations": [],
        // Specify workspace-specific launch compounds
        "compounds": []
      }
    }
    ```

1. Select the **Open Workspace** button to view your new workspace.
  
In this workspace, you will have access to launch configurations, Jest test suites, code search, file search, and many other features.

## Wrapping up

> While these tools aren't a replacement for a thorough understanding of the OSD codebase, they help streamline your development workflow. By automating tasks like server startup, debugging, and linting, they reduce time spent on configuration and allow you to focus on writing code. These tools make the development and PR process more efficient, saving you time and boosting productivity.

Reviewer 2: ```suggestion
While these tools aren't a replacement for a thorough understanding of the OSD codebase, they can help to streamline your development workflow. By automating tasks like server startup, debugging, and linting, they reduce time spent on configuration and allow you to focus on writing code. These tools make the development and PR process more efficient, saving you time and boosting productivity.
```

---
layout: post
title:  "Streamline OpenSearch Dashboards development with VS Code"
authors:
 - wronghuy
 - kolchfa

> date: 2024-12-17

Reviewer: Please update the publish date to 2024-12-26

> date: 2024-12-17

Reviewer 3: ```suggestion
date: 2024-12-27
```
